puzzles/puzzles.go

```diff
@@ -124,6 +135,91 @@ func CELOnlyPuzzle(g *game.Game, moves []*move.Move) (bool, pb.PuzzleTag) {
 	return isCEL, pb.PuzzleTag_CEL_ONLY
 }
 
+func tagsFitInBucket(tags []pb.PuzzleTag, bucket *pb.PuzzleBucket) bool {
+	tagMap := map[pb.PuzzleTag]bool{}
+	for _, tag := range tags {
+		tagMap[tag] = true
+	}
+
+	excludesMap := map[pb.PuzzleTag]bool{}
+	for _, tag := range bucket.Excludes {
+		excludesMap[tag] = true
+	}
+
+	for _, tag := range tags {
+		if excludesMap[tag] {
+			return false
+		}
+	}
+
+	for _, includeTag := range bucket.Includes {
+		if !tagMap[includeTag] {
+			return false
+		}
+	}
+	return true
+}
+
+func validatePuzzleGenerationRequest(req *pb.PuzzleGenerationRequest) error {
+	if req == nil {
+		return errors.New("puzzle generation request is nil")
+	}
+	if req.Buckets == nil {
+		return errors.New("buckets are nil in puzzle generation request")
+	}
+	bucketEncryptions := map[int]bool{}
+	for idx, bucket := range req.Buckets {
+		be, err := validatePuzzleBucket(bucket)
+		if err != nil {
+			return fmt.Errorf("error for bucket %d: %s", idx, err.Error())
+		}
+		if bucketEncryptions[be] {
```

Comment:
should you modify `bucketEncryptions` in this loop as well?

Reply:
I don't understand. This is just a validation. What would we need to modify?

Reply:
i have no idea what this was. i think you can ignore it 😅

Reply:
i have no idea what this was. i think you can ignore it 😅